### PR TITLE
hifiasm trio assembly

### PR DESCRIPTION
**Pull request by juniper-lake**

### CODE REVIEW

File: process_cohort.smk [hunk 1]
```diff
@@ -42,6 +58,36 @@ else:
...
+ubam_dict = defaultdict(dict)
+fastq_pattern = re.compile(r'smrtcells/ready/(?P<sample>[A-Za-z0-9_-]+)/(?P<movie>m\d{5}[Ue]?_\d{6}_\d{6}).fastq.gz')
+fastq_dict = defaultdict(dict)
+ubam_fastq_dict = defaultdict(dict)
+for infile in Path('smrtcells/ready').glob('**/*.bam'):
+    ubam_match = ubam_pattern.search(str(infile))
+    if ubam_match:
+        # create a dict-of-dict to link samples to movie context to uBAM filenames
+        ubam_dict[ubam_match.group('sample')][ubam_match.group('movie')] = str(infile)
+        ubam_fastq_dict[ubam_match.group('sample')][ubam_match.group('movie')] = str(infile)
```
Comment (williamrowell) [inline process_cohort.smk:82]: What happens if both a uBAM and a FASTQ are available for the same movie?

Comment (williamrowell) [reply process_cohort.smk:82]: Looks like the answer is, maybe nothing?  If you only need the keys from `ubam_fastq_dict[sample]`, might be able to get them from `movie_dict` instead.

File: process_cohort.smk [hunk 2]
```diff
@@ -74,6 +120,19 @@ else:
 targets = []
 include: 'rules/cohort_common.smk'
 
+# assemble with hifiasm
+include: 'rules/cohort_hifiasm.smk'
+if 'trio_assembly' in config['cohort_targets']:
+    # assembly and stats
+    targets.extend([f"cohorts/{cohort}/hifiasm/{trio}.asm.dip.{infix}.{suffix}"
+                for suffix in ['fasta.gz', 'fasta.stats.txt']
+                for infix in ['hap1.p_ctg', 'hap2.p_ctg']
```
Comment (amwenger) [inline process_cohort.smk:129]: Perhaps add the parent ID to the output files too to clarify which parent is hap1 and which parent is hap2

File: process_cohort.smk (hunk 1 above)
Comment (juniper-lake) [reply process_cohort.smk:82]: Good point. I was trying to avoid dependency on outputs from process_sample

File: process_cohort.smk (hunk 2 above)
Comment (juniper-lake) [reply process_cohort.smk:129]: Good idea

File: process_cohort.smk (hunk 1 above)
Comment (williamrowell) [reply process_cohort.smk:82]: Could probably grab the keys from `ubam_dict[sample]` and `fastq_dict[sample]` and merge them.

File: rules/cohort_hifiasm.smk [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    input: f"cohorts/{cohort}/hifiasm/{{sample}}.asm.dip.{{infix}}.fasta.gz"
+    output: f"cohorts/{cohort}/hifiasm/{{sample}}.asm.dip.{{infix}}.fasta.stats.txt"
+    log: f"cohorts/{cohort}/logs/asm_stats/{{sample}}.asm.dip.{{infix}}.fasta.log"
+    benchmark: f"cohorts/{cohort}/benchmarks/asm_stats/{{sample}}.asm.dip.{{infix}}.fasta.tsv"
+    conda: "envs/k8.yaml"
+    message: "Executing {rule}: Calculating stats for {input}."
+    shell: f"(k8 workflow/scripts/calN50/calN50.js -f {config['ref']['index']} {{input}} > {{output}}) > {{log}} 2>&1"
+
+
+rule align_hifiasm:
```
Comment (williamrowell) [inline rules/cohort_hifiasm.smk:98]: We chatted about this offline, but might be nice in a future release to align each haplotype separately and use the parent name as part of the read group, then merge the results.

File: process_cohort.smk (hunk 2 above)
Comment (juniper-lake) [reply process_cohort.smk:127]: Added a haplotype key file to the output since hifiasm doesn't seem to have an option for haplotype-specific file prefixes

**Comment by amwenger** [comment]
Impressive!  Looks really good.

**Review by williamrowell — APPROVED** [review]
LGTM

**Pull request merged**